### PR TITLE
SSD classifier

### PR DESCRIPTION
**Pull request by Elgeweily**

### CODE REVIEW

**Comment by ser94mor** [comment]
This pull request **introduces 1 alert** when merging 9a6cef5651becfdef10b28c2ecd7211232359626 into ed7da4b3a9eccae23995ba40593c626d1bc79cd2 - [view on LGTM.com](https://lgtm.com/projects/g/aoool/self-driving-car-using-ros/rev/pr-eaa918e0f59655509b4252dd33b504f488940feb)

**new alerts:**

* 1 for Unused local variable

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: ros/src/tl_detector/light_classification/tl_classifier.py [hunk 1]
```diff
@@ -2,10 +2,30 @@
 import cv2
 import numpy as np
 
+import tensorflow as tf
+from helpers import load_graph
+
+import os.path as path
+
+# Load frozen graph of trained model
+five_up = path.abspath(path.join(__file__, '../../../../..'))
```
Comment (Karthikeya108) [inline ros/src/tl_detector/light_classification/tl_classifier.py]: I suggest the following option to load the model.
1. Store the model under tl_detector directory.
2. Dynamically fetch the path 
`import rospkg`
`base_path = r.get_path('tl_detector')`
`SSD_INCEPTION_SIM = path.join(base_path,'models/ssd_inception_v2_alex_sim/frozen_inference_graph.pb')`

File: ros/src/tl_detector/light_classification/tl_classifier.py [hunk 2]
```diff
@@ -2,10 +2,30 @@
...
+# Each score represent the level of confidence for each of the objects
+detection_scores = detection_graph.get_tensor_by_name('detection_scores:0')
+# The classification of the object (integer id)
+detection_classes = detection_graph.get_tensor_by_name('detection_classes:0')
+
 class TLClassifier(object):
     def __init__(self):
-        #TODO load classifier
-        pass
+	tf.reset_default_graph()
```
Comment (Karthikeya108) [inline ros/src/tl_detector/light_classification/tl_classifier.py]: Suggestion:
Load the tf model within the init file

File: ros/src/tl_detector/light_classification/tl_classifier.py [hunk 3]
```diff
@@ -35,13 +55,28 @@ def simple_opencv_red_color_classifier(self,image):
...
+        (boxes, scores, classes) = self.sess.run([detection_boxes, detection_scores, detection_classes], feed_dict={image_tensor: image_np})
+
+        # Remove unnecessary dimensions
+        scores = np.squeeze(scores)
+        classes = np.squeeze(classes)
+        
+        for i in range(len(classes)):
+		print('class=', classes[i])
+		print('score=', scores[i])
+            	if classes[i] == 2 and scores[i] > 0.5: # if red light with confidence more than 50%
```
Comment (Karthikeya108) [inline ros/src/tl_detector/light_classification/tl_classifier.py]: Suggestion:
Also return the Yellow light, (although we ignore the yellow light for now, might be useful later)

**Review by Karthikeya108 — CHANGES_REQUESTED** [review]

File: ros/src/tl_detector/light_classification/tl_classifier.py (hunk 1 above)
Comment (thejerrytan) [reply ros/src/tl_detector/light_classification/tl_classifier.py]: Yes, i agree with @Karthikeya108  on this, lets all agree to keep all our models in a models folder in the light_classification directory, since @Karthikeya108  also kept his model there. so the suggested structure would be:

<pre>
  /light_classification
    /models
        /ssd_inception_v2_coco_sim
             - graph_inference.pb
        /simple_cnn_classifier_sim
             - model.h5
    - tl_classifier.py
</pre>

File: ros/src/tl_detector/light_classification/tl_classifier.py (hunk 3 above)
Comment (thejerrytan) [reply ros/src/tl_detector/light_classification/tl_classifier.py]: can you print the name of the classes? i.e. 0 == green, 1 == yellow 2 == red, 3 == unknown?

File: ros/src/tl_detector/light_classification/tl_classifier.py (hunk 2 above)
Comment (thejerrytan) [reply ros/src/tl_detector/light_classification/tl_classifier.py]: yes

File: ros/src/tl_detector/light_classification/tl_classifier.py (hunk 3 above)
Comment (ser94mor) [reply ros/src/tl_detector/light_classification/tl_classifier.py]: I think we should treat yellow light as red. When yellow follows green, the driver should start decelerating. When yellow follows red, the driver should prepare to accelerate but not start accelerating until there is a green light. For an unsophisticated self-driving car, yellow essentially means red.

Comment (Elgeweily) [reply ros/src/tl_detector/light_classification/tl_classifier.py]: Makes sense, I will change it.....also I made the threshold for stopping 1 instance of red state, but the threshold for driving from a stop, 3 persisting instances of non-red state, which makes it more robust.

File: ros/src/tl_detector/light_classification/tl_classifier.py (hunk 1 above)
Comment (Elgeweily) [reply ros/src/tl_detector/light_classification/tl_classifier.py]: Done

File: ros/src/tl_detector/light_classification/tl_classifier.py (hunk 2 above)
Comment (Elgeweily) [reply ros/src/tl_detector/light_classification/tl_classifier.py]: Done.

File: ros/src/tl_detector/tl_detector.py [hunk 4]
```diff
@@ -93,6 +94,13 @@ def image_cb(self, msg):
         :param msg: image from car-mounted camera
         :type msg: Image
         """
+        if self.last_state == TrafficLight.RED:
```
Comment (thejerrytan) [inline ros/src/tl_detector/tl_detector.py]: please move the chunk from line 52 - 56 to before the subscriber gets initilaized in the __init__ function to prevent a race condition! thanks!

File: ros/src/tl_detector/tl_detector.py [hunk 5]
```diff
@@ -93,6 +94,13 @@ def image_cb(self, msg):
...
+            # if stopped, start driving only after detecting consistent non-red
+            STATE_COUNT_THRESHOLD = STATE_COUNT_THRESHOLD_DRIVE
+	else:	
+            # Stop on detecting a single instance of red
+            STATE_COUNT_THRESHOLD = STATE_COUNT_THRESHOLD_STOP
```
Comment (thejerrytan) [inline ros/src/tl_detector/tl_detector.py]: generally variables in CAPS in python are constants, shouldnt be changed. 

Also, you are assuming when the last state is RED, the car is stopped. It could have detected the red lights from far away. Would that affect the program in any unforseen manner?

File: ros/src/tl_detector/light_classification/tl_classifier.py [hunk 6]
```diff
@@ -35,13 +58,28 @@ def simple_opencv_red_color_classifier(self,image):
...
+        (boxes, scores, classes) = self.sess.run([self.detection_boxes, self.detection_scores, self.detection_classes], feed_dict={self.image_tensor: image_np})
+
+        # Remove unnecessary dimensions
+        scores = np.squeeze(scores)
+        classes = np.squeeze(classes)
+        
+        for i in range(len(classes)):
+	    print('class =', labels_dict[classes[i]])
+	    print('score =', scores[i])
+            if (classes[i] == 2 or classes[i] == 3) and scores[i] > 0.1: # if red or yellow light with confidence more than 10%
```
Comment (thejerrytan) [inline ros/src/tl_detector/light_classification/tl_classifier.py]: if confidence threshold is so low for this to work, usually it means that our model is not accurate enough or not trained enough yet - not enough epochs or too small dataset.

**Review by thejerrytan — CHANGES_REQUESTED** [review]

**Comment by thejerrytan** [comment]
pls do a quick check, convert tabs to spaces for indentation. I am not sure why i have tabs as indentation when i pull this branch

**Comment by thejerrytan** [comment]
https://carnd.slack.com/files/U9AATNYCT/FFLCQ27E0/ssd_classifier.mp4

File: ros/src/tl_detector/light_classification/tl_classifier.py (hunk 6 above)
Comment (Elgeweily) [reply ros/src/tl_detector/light_classification/tl_classifier.py]: Yeah you are right, but the point is, it's low only from far away, and the car needs to prepare and start braking early enough, that's why it should start braking at the slightest hint of red....but when you get near the light, the accuracy shoots up to 99%. But of course, if we have more data, it will detect even the far away lights with high confidence.

File: ros/src/tl_detector/tl_detector.py (hunk 5 above)
Comment (Elgeweily) [reply ros/src/tl_detector/tl_detector.py]: No, it's good, even if the car sees red while not stopped, it will be slowing down, so it should not start accelerating again, until it sees 3 repeated instances of green, so the comment should be: "# if stopped or slowing down, start accelerating only after detecting consistent green (or consistent unknown)"

File: ros/src/tl_detector/tl_detector.py (hunk 4 above)
Comment (Elgeweily) [reply ros/src/tl_detector/tl_detector.py]: It's the same in the master branch, is it possible to request a change in the master branch?

File: ros/src/tl_detector/tl_detector.py (hunk 5 above)
Comment (Elgeweily) [reply ros/src/tl_detector/tl_detector.py]: Yeah you are right, but the point is, it's low only from far away, and the car needs to prepare and start braking early enough, that's why it should start braking at the slightest hint of red....but when you get near the light, the accuracy shoots up to 99%. But of course, if we have more data, it will detect even the far away lights with high confidence.

File: ros/src/tl_detector/light_classification/tl_classifier.py (hunk 6 above)
Comment (Elgeweily) [reply ros/src/tl_detector/light_classification/tl_classifier.py]: Yeah you are right, but the point is, it's low only from far away, and the car needs to prepare and start braking early enough, that's why it should start braking at the slightest hint of red....but when you get near the light, the accuracy shoots up to 99%. But of course, if we have more data, it will detect even the far away lights with high confidence.

File: ros/src/tl_detector/tl_detector.py (hunk 5 above)
Comment (Elgeweily) [reply ros/src/tl_detector/tl_detector.py]: - yeah I will change the all caps variable.

- No, it's good, even if the car is not stopped, it will be slowing down, so it should not start accelerating again until it sees 3 repeated instances of non-red, so the comment should be: "# if stopped or slowing down, start accelerating only after detecting consistent non-red (yellow considered red)

**Comment by thejerrytan** [comment]
> This pull request **introduces 1 alert** when merging [9a6cef5](https://github.com/aoool/self-driving-car-using-ros/commit/9a6cef5651becfdef10b28c2ecd7211232359626) into [ed7da4b](https://github.com/aoool/self-driving-car-using-ros/commit/ed7da4b3a9eccae23995ba40593c626d1bc79cd2) - [view on LGTM.com](https://lgtm.com/projects/g/aoool/self-driving-car-using-ros/rev/pr-eaa918e0f59655509b4252dd33b504f488940feb)
> 
> **new alerts:**
> 
> * 1 for Unused local variable
> 
> _Comment posted by [LGTM.com](https://lgtm.com)_

think its fine, we might need the bounding boxes in future.

**Comment by Elgeweily** [comment]
> pls do a quick check, convert tabs to spaces for indentation. I am not sure why i have tabs as indentation when i pull this branch

**Comment by ser94mor** [comment]
This pull request **introduces 3 alerts** when merging 680555e6d8b17726da0190a6602da46e9d162608 into ed7da4b3a9eccae23995ba40593c626d1bc79cd2 - [view on LGTM.com](https://lgtm.com/projects/g/aoool/self-driving-car-using-ros/rev/pr-7a58307554a43f26285d30064036e6f0c591b827)

**new alerts:**

* 3 for Unused local variable

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by ser94mor** [comment]
This pull request **introduces 1 alert** when merging e333eb88afb470bef28757d14e3653ee60f77d57 into ed7da4b3a9eccae23995ba40593c626d1bc79cd2 - [view on LGTM.com](https://lgtm.com/projects/g/aoool/self-driving-car-using-ros/rev/pr-75b15f516954b0d5f75662a56917bbd8d8769d1d)

**new alerts:**

* 1 for Unused local variable

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Review by Karthikeya108 — APPROVED** [review]

**Review by ser94mor — CHANGES_REQUESTED** [review]
I tested it for the simulator. It still detects red traffic lights too late, does not stop in front of the stop line, and continue driving. @pkorivi's classifier is good for the simulator and is very resource-friendly. I suggest concentrating on the real-world images and apply the SSD classifier to the available videos.

File: ros/src/tl_detector/light_classification/tl_classifier.py [hunk 7]
```diff
@@ -2,10 +2,33 @@
...
+import rospkg
+import os.path as path
+
+# Model path
+r = rospkg.RosPack()
+base_path = r.get_path('tl_detector')
+SSD_INCEPTION_SIM = path.join(base_path,'light_classification/models/ssd_inception_v2_coco_retrained_sim/frozen_inference_graph.pb')
+
+# Labels dictionary
+labels_dict = {1: 'Green', 2: 'Red', 3: 'Yellow',4: 'Unknown'}
```
Comment (ser94mor) [inline ros/src/tl_detector/light_classification/tl_classifier.py:18]: Does `4: "Unknown"` mean that there were no traffic lights detected or you've had a special class "Unknown" during training. I suggest avoiding using "Unknown" class for the training, it does not make sense for the neural network without a huge variety of examples from the class "Unknown."

File: ros/src/tl_detector/light_classification/tl_classifier.py [hunk 8]
```diff
@@ -2,10 +2,33 @@
...
+
 class TLClassifier(object):
     def __init__(self):
-        #TODO load classifier
-        pass
+        # Load frozen graph of trained model
```
Comment (ser94mor) [inline ros/src/tl_detector/light_classification/tl_classifier.py:22]: This implementation does not leave together with the OpenCV-based classifier. Consider creating a subclass of the TLClassifier for your case (as well as for OpenCV-based). TLClassifier, in this case, can be an abstract class.

TLDetector, in this case, should invoke a [factory method](https://en.wikipedia.org/wiki/Factory_method_pattern) to create a corresponding TLClassifier instance.

File: ros/src/tl_detector/light_classification/tl_classifier.py [hunk 9]
```diff
@@ -35,13 +58,28 @@ def simple_opencv_red_color_classifier(self,image):
...
+        # Actual detection
+        (boxes, scores, classes) = self.sess.run([self.detection_boxes, self.detection_scores, self.detection_classes], feed_dict={self.image_tensor: image_np})
+
+        # Remove unnecessary dimensions
+	boxes = np.squeeze(boxes)
+        scores = np.squeeze(scores)
+        classes = np.squeeze(classes)
+
+        for i in range(len(classes)):
+            print('class =', labels_dict[classes[i]])
```
Comment (ser94mor) [inline ros/src/tl_detector/light_classification/tl_classifier.py]: Please, use ROS logging facilities instead of using `print()`

File: ros/src/tl_detector/light_classification/tl_classifier.py (hunk 6 above)
Comment (ser94mor) [reply ros/src/tl_detector/light_classification/tl_classifier.py]: You can detect objects that are far away if you increase the input image resolution. Can SSD handle such an increase?

File: ros/src/tl_detector/light_classification/tl_classifier.py [hunk 10]
```diff
@@ -35,13 +58,28 @@ def simple_opencv_red_color_classifier(self,image):
...
+            print('score =', scores[i])
+            if (classes[i] == 2 or classes[i] == 3) and scores[i] > 0.1: # if red or yellow light with confidence more than 10%
+                return TrafficLight.RED
```
Comment (ser94mor) [inline ros/src/tl_detector/light_classification/tl_classifier.py:76]: Red lights should be classified as red, yellow traffic lights should be classified as yellow. TLDetector should treat red and yellow traffic lights as the same, not the TLClassifier.

File: ros/src/tl_detector/tl_detector.py (hunk 4 above)
Comment (ser94mor) [reply ros/src/tl_detector/tl_detector.py]: Yes, let us do it in a separate bugfix PR to a master branch.

**Review by ser94mor — CHANGES_REQUESTED** [review]

File: ros/src/tl_detector/light_classification/tl_classifier.py (hunk 7 above)
Comment (Elgeweily) [reply ros/src/tl_detector/light_classification/tl_classifier.py:18]: It is an Unknown class.......yeah right, the dataset doesn't contain a single instance of Unknown, so I can retrain on 3 classes only

**Review by thejerrytan — APPROVED** [review]

File: ros/src/tl_detector/light_classification/tl_classifier.py (hunk 9 above)
Comment (Elgeweily) [reply ros/src/tl_detector/light_classification/tl_classifier.py]: Done.

File: ros/src/tl_detector/light_classification/tl_classifier.py (hunk 8 above)
Comment (Elgeweily) [reply ros/src/tl_detector/light_classification/tl_classifier.py:22]: Awaiting code refactoring.

Comment (pkorivi) [reply ros/src/tl_detector/light_classification/tl_classifier.py:22]: I merged the changes which I made yesterday.

File: ros/src/tl_detector/light_classification/tl_classifier.py (hunk 10 above)
Comment (Elgeweily) [reply ros/src/tl_detector/light_classification/tl_classifier.py:76]: Done.
